File: src/test/java/uk/gov/justice/hmpps/prison/service/PrisonerCreationServiceTest.java
package uk.gov.justice.hmpps.prison.service;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import uk.gov.justice.hmpps.prison.repository.PrisonerRepository;
import uk.gov.justice.hmpps.prison.repository.jpa.model.NomsIdSequence;
import uk.gov.justice.hmpps.prison.service.transformers.OffenderTransformer;

import java.time.Clock;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class PrisonerCreationServiceTest {
    @Mock
    private PrisonerRepository prisonerRepository;

    private PrisonerCreationService service;
    public static final NomsIdSequence START_SEQUENCE = NomsIdSequence.builder()
            .suffixAlphaSeq(27)
            .currentSuffix("AA")
            .currentPrefix("A")
            .prefixAlphaSeq(1)
            .nomsId(0)
            .build();

    @BeforeEach
    void setUp() {
        service = new PrisonerCreationService(prisonerRepository, null, null, null, null, null, null, new OffenderTransformer(Clock.systemUTC()));
    }

    @Test
    void testNextSeqenceCreated() {
        when(prisonerRepository.getNomsIdSequence()).thenReturn(START_SEQUENCE);
        final var nextSeq = START_SEQUENCE.next();
        when(prisonerRepository.updateNomsIdSequence(Mockito.eq(nextSeq), Mockito.eq(START_SEQUENCE))).thenReturn(1);

        final var nextPrisonerIdentifier = service.getNextPrisonerIdentifier();

        assertThat(nextPrisonerIdentifier.getId()).isEqualTo(START_SEQUENCE.getPrisonerIdentifier());
    }

    @Test
    void testCannotUpdate() {
        when(prisonerRepository.getNomsIdSequence()).thenReturn(START_SEQUENCE);
        final var nextSeq = START_SEQUENCE.next();
        when(prisonerRepository.updateNomsIdSequence(Mockito.eq(nextSeq), Mockito.eq(START_SEQUENCE))).thenReturn(0);

        assertThatThrownBy(() -> service.getNextPrisonerIdentifier())
                .hasMessage("Prisoner Identifier cannot be generated, please try again");
    }

    @Test
    void testCanUpdateAfterRetry() {
        final var nextSeq = START_SEQUENCE.next();
        when(prisonerRepository.getNomsIdSequence()).thenReturn(START_SEQUENCE, nextSeq);
        when(prisonerRepository.updateNomsIdSequence(Mockito.eq(nextSeq), Mockito.eq(START_SEQUENCE))).thenReturn(0);
        final var nextNextSeq = nextSeq.next();
        when(prisonerRepository.updateNomsIdSequence(Mockito.eq(nextNextSeq), Mockito.eq(nextSeq))).thenReturn(1);

        final var nextPrisonerIdentifier = service.getNextPrisonerIdentifier();

        assertThat(nextPrisonerIdentifier.getId()).isEqualTo(nextSeq.getPrisonerIdentifier());
    }
}

File: src/test/java/uk/gov/justice/hmpps/prison/service/enteringandleaving/PrisonerCreationServiceTest.kt
package uk.gov.justice.hmpps.prison.service.enteringandleaving

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.mockito.Mock
import org.mockito.Mockito.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import uk.gov.justice.hmpps.prison.repository.PrisonerRepository
import uk.gov.justice.hmpps.prison.repository.jpa.model.NomsIdSequence

class PrisonerCreationServiceTest {
  @Mock
  private val prisonerRepository: PrisonerRepository = mock()

  private lateinit var service: PrisonerCreationService

  companion object {
    val START_SEQUENCE = NomsIdSequence(
      0,
      1,
      27,
      "A",
      "AA",
    )
  }

  @BeforeEach
  internal fun setUp() {
    service = PrisonerCreationService(
      genderRepository = mock(),
      ethnicityRepository = mock(),
      titleRepository = mock(),
      suffixRepository = mock(),
      offenderIdentifierRepository = mock(),
      offenderRepository = mock(),
      offenderTransformer = mock(),
      prisonerRepository = prisonerRepository,
      bookingIntoPrisonService = mock(),
    )
  }

  @Nested
  @DisplayName("getNextPrisonerIdentifier")
  inner class GetNextPrisonerIdentifier {
    @Test
    fun `next sequence is created`() {
      whenever(prisonerRepository.nomsIdSequence).thenReturn(START_SEQUENCE)
      val nextSeq = START_SEQUENCE.next()
      whenever(prisonerRepository.updateNomsIdSequence(eq(nextSeq), eq(START_SEQUENCE))).thenReturn(1)
      val nextPrisonerIdentifier = service.getNextPrisonerIdentifier()
      assertThat(nextPrisonerIdentifier.id).isEqualTo(START_SEQUENCE.prisonerIdentifier)
    }

    @Test
    fun `exception thrown when can never get next sequence`() {
      whenever(prisonerRepository.nomsIdSequence).thenReturn(START_SEQUENCE)
      val nextSeq = START_SEQUENCE.next()
      whenever(prisonerRepository.updateNomsIdSequence(eq(nextSeq), eq(START_SEQUENCE))).thenReturn(0)

      assertThatThrownBy { service.getNextPrisonerIdentifier() }
        .hasMessage("Prisoner Identifier cannot be generated, please try again")
    }

    @Test
    fun `will try again when initially get not get a next sequence`() {
      val nextSeq = START_SEQUENCE.next()
      whenever(prisonerRepository.nomsIdSequence).thenReturn(START_SEQUENCE, nextSeq)
      whenever(prisonerRepository.updateNomsIdSequence(eq(nextSeq), eq(START_SEQUENCE))).thenReturn(0)
      val nextNextSeq = nextSeq.next()

      whenever(prisonerRepository.updateNomsIdSequence(eq(nextNextSeq), eq(nextSeq))).thenReturn(1)
      val nextPrisonerIdentifier = service.getNextPrisonerIdentifier()

      assertThat(nextPrisonerIdentifier.id).isEqualTo(nextSeq.prisonerIdentifier)
    }
  }
}
